experimental/zerotrie/src/reader.rs

```diff
@@ -0,0 +1,454 @@
+        None => trie.len(),
+    };
+    debug_get_range(trie, p..q)
+}
+
+enum ByteType {
+    Ascii,
+    Span,
+    Value,
+    Match,
+}
+
+impl core::fmt::Debug for ByteType {
+    fn fmt(&self, f: &mut core::fmt::Formatter) -> core::fmt::Result {
+        use ByteType::*;
+        f.write_str(match *self {
+            Ascii => "a",
+            Span => "s",
+            Value => "v",
+            Match => "m",
+        })
+    }
+}
+
+#[inline]
+fn byte_type(b: u8) -> ByteType {
+    match b & 0b11100000 {
+        0b10000000 => ByteType::Value,
+        0b10100000 => ByteType::Span,
+        0b11000000 => ByteType::Match,
+        0b11100000 => ByteType::Match,
+        _ => ByteType::Ascii,
+    }
+}
+
+// DISCUSS: This function is 7% faster *on aarch64* if we assert a max on w.
+//
+// | Bench         | No Assert, x86_64 | No Assert, aarch64 | Assertion, x86_64 | Assertion, aarch64 |
+// |---------------|-------------------|--------------------|-------------------|--------------------|
+// | basic         | ~187.51 ns        | ~97.586 ns         | ~199.11 ns        | ~99.236 ns         |
+// | subtags_10pct | ~9.5557 µs        | ~4.8696 µs         | ~9.5779 µs        | ~4.5649 µs         |
+// | subtags_full  | ~137.75 µs        | ~76.016 µs         | ~142.02 µs        | ~70.254 µs         |
+
+/// Query the trie assuming all branch nodes are binary search.
+pub fn get_bsearch_only(mut trie: &[u8], mut ascii: &[u8]) -> Option<usize> {
+    loop {
+        let (b, x, i, search);
+        (b, trie) = trie.split_first()?;
+        let byte_type = byte_type(*b);
+        (x, trie) = match byte_type {
+            ByteType::Ascii => (0, trie),
+            ByteType::Span | ByteType::Value => read_extended_varint(*b, trie)?,
+            ByteType::Match => read_varint(*b, trie)?,
+        };
+        if let Some((c, temp)) = ascii.split_first() {
+            if matches!(byte_type, ByteType::Ascii) {
+                if b == c {
+                    // Matched a byte
+                    ascii = temp;
+                    continue;
+                } else {
+                    // Byte that doesn't match
+                    return None;
+                }
+            }
+            if matches!(byte_type, ByteType::Value) {
+                // Value node, but not at end of string
+                continue;
+            }
+            if matches!(byte_type, ByteType::Span) {
+                let (trie_span, ascii_span);
+                (trie_span, trie) = debug_split_at(trie, x)?;
+                (ascii_span, ascii) = maybe_split_at(ascii, x)?;
+                if trie_span == ascii_span {
+                    // Matched a byte span
+                    continue;
+                } else {
+                    // Byte span that doesn't match
+                    return None;
+                }
+            }
+            // Branch node
+            let (x, w) = if x >= 256 { (x & 0xff, x >> 8) } else { (x, 0) };
```

Comment:
suggestion: you _may_ be able to play around with how you're doing this calculation to get the same perf win as the assertion

Reply:
Happy to hear suggestions. I tried several formulations of this calculation and this was the best of the ones I tried, but there could be more formulations I didn't try. Might be easier once the code is landed.